File: arduino/WUS_Control/TUSS4470.cpp
/*
   TUSS4470.cpp - Library for controlling ultrasound burst excitation with TUSS4470. Listening is not supported.
   Author: Eric Hu, 1/10/2022
*/

#include "TUSS4470.h"

/*
   Create a TUSS4470 object
*/
TUSS4470 :: TUSS4470(uint8_t CSpin, bool pre_driver) {
  // Chip select pin to enable SPI (active low)
  pinMode(CSpin, OUTPUT);
  digitalWrite(CSpin, HIGH);
  
  _CSpin = CSpin;
  _drvVoltage = 0;
  _drvCurrent = LOW_CURRENT;
  _drvHiZ = HIZ_ON;
  _pulseCount = 0;
  _ioMode = IO_MODE0;
  _preDriver = pre_driver;
}


// ----------------------------- General functions ----------------------------- //

void TUSS4470 :: begin() {
//  SPI.begin();
  delay(100);
}

void TUSS4470 :: reset() {
  for (int i = 0; i < 10; i++) {
    controlRegister(WRITE, RESET_ADDS[i], RESET_DATA[i]);
  }
  _drvVoltage = 0;
  _drvCurrent = LOW_CURRENT;
  _drvHiZ = HIZ_ON;
  _pulseCount = 0;
  _ioMode = IO_MODE0;
}

void TUSS4470 :: disableRegulation() {
  _drvHiZ = HIZ_ON;
  updateVRDVRegister();
}

void TUSS4470 :: enableRegulation() {
  _drvHiZ = HIZ_OFF;
  updateVRDVRegister();
}

void TUSS4470 :: disablePreDriver() {
  _preDriver = false;
  updateBurstPulseRegister();
}

void TUSS4470 :: enablePreDriver() {
  _preDriver = true;
  updateBurstPulseRegister();
}

void TUSS4470 :: set(uint8_t voltage, CurrentType current, IOMode mode, uint8_t pulse_count) {
  setVoltage(voltage);
  setCurrent(current);
  setIOMode(mode);
  setPulseCount(pulse_count);
}

void TUSS4470 :: setVoltage(uint8_t volt) {
  volt = (volt < 5) ? 0 : (volt - 5); // Set to 0 if under 5
  volt &= 0xF;  // Only count last 4 bits
  _drvVoltage = volt;
  updateVRDVRegister();
}

void TUSS4470 :: setCurrent(CurrentType curr) {
  _drvCurrent = curr;
  updateVRDVRegister();
}

void TUSS4470 :: setIOMode(IOMode mode) {
  _ioMode = mode;
  updateDevConfigRegister();
}

void TUSS4470 :: setPulseCount(uint8_t count) {
  count &= 0x3F; // Only count last 6 bits
  _pulseCount = count;
  updateBurstPulseRegister();
}


// ----------------------------- Private functions ----------------------------- //

void TUSS4470 :: updateVRDVRegister() {
  Serial.println("VDRV");
  uint8_t data = (_drvHiZ << 5) + (_drvCurrent << 4) + _drvVoltage;
  controlRegister(WRITE, 0x16, data);
}

void TUSS4470 :: updateBurstPulseRegister() {
  Serial.println("BurstPulse");
  uint8_t data = _pulseCount;
  if (_preDriver) {data += 0x40;}
  controlRegister(WRITE, 0x1A, data);
}

void TUSS4470 :: updateDevConfigRegister() {
  Serial.println("DevConfig");
  uint8_t data = _ioMode;
  controlRegister(WRITE, 0x14, data);
}

void TUSS4470 :: controlRegister(CommandType rw, byte address, byte data) {
  address &= 0x3F;
  uint16_t comm = rw;
  comm += ((uint16_t)address << 9);
  comm += data;
  if (!findParity16(comm)) {comm += 0x0100;}

  transferCommand(comm);
}

bool TUSS4470 :: findParity16(uint16_t num) {
  num ^= num >> 8;
  num ^= num >> 4;
  num ^= num >> 2; 
  num ^= num >> 1;
  return num & 1;
}

void TUSS4470 :: transferCommand(uint16_t data16) {
  SPI.beginTransaction(SPISettings(20000000, MSBFIRST, SPI_MODE1));
  digitalWrite(_CSpin, LOW);
  uint16_t rec16 = SPI.transfer16(data16);
  digitalWrite(_CSpin, HIGH);
  Serial.println(data16, BIN);
  Serial.println(rec16, BIN);
  Serial.println();
}

File: lib/wus_control/hardware/TUSS4470.py
# TUSS4470.py - Library for controlling ultrasound burst excitation with TUSS4470. Listening is not supported.
# Author: Eric Hu, 6/14/2022

from adafruit_bus_device.spi_device import SPIDevice

RESET_DICT = {0x10:0x00, 0x11:0x00, 0x12:0x00, 0x13:0x00, 0x14:0x00, 0x16:0x20, 0x17:0x07, 0x18:0x14, 0x1A:0x00, 0x1B:0x00}
RW = {'READ': 0x8000, 'WRITE': 0x0000}
CURRENT = {'LOW': 0, 'HIGH': 1}
HIZ = {'OFF': 0, 'ON': 1}
IO = {'MODE0': 0, 'MODE1': 1, 'MODE2': 2, 'MODE3': 3}
PRE_DRIVER = {'OFF': 0, 'ON': 1}

# Create a TUSS4470 object
class TUSS4470(object):

    def __init__(self, cs_pin):
        # Chip select pin to enable SPI (active low)
        self._cs = cs_pin
        self._device = None

        self._voltage = None
        self._pulse_count = None
        self._hi_current_mode = CURRENT['LOW']
        self._hi_z_mode = HIZ['ON']
        self._io_mode = IO['MODE0']
        self._pre_driver_mode = PRE_DRIVER['OFF']

# ----------------------------- General functions ----------------------------- #

    def begin(self, spi_port):
        self._device = SPIDevice(spi=spi_port, chip_select=self._cs, baudrate=500000, polarity=0, phase=1)

    def reset(self):
        for key in RESET_DICT:
            self.control_register(RW['WRITE'], key, RESET_DICT[key]);
        self._voltage = 0;
        self._pulse_count = 0
        self._hi_current_mode = CURRENT['LOW']
        self._hi_z_mode = HIZ['ON']
        self._io_mode = IO['MODE0']
        self._pre_driver_mode = PRE_DRIVER['OFF']

    def disableRegulation(self):
        self._hi_z_mode = HIZ['ON']
        self._update_VRDV_register()

    def enableRegulation(self):
        self._hi_z_mode = HIZ['OFF']
        self._update_VRDV_register()

    def disablePreDriver(self):
        self._pre_driver_mode = PRE_DRIVER['OFF'];
        self._update_BP_register()

    def enablePreDriver(self):
        self._pre_driver_mode = PRE_DRIVER['ON'];
        self._update_BP_register()

    def set(self, voltage, current, io_mode, pulse_count):
        self.set_voltage(voltage);
        self.set_current(current);
        self.set_io_mode(io_mode);
        self.set_pulse_count(pulse_count);

    def set_voltage(self, voltage):
        voltage = max(voltage - 5, 0)   # Voltage must be at least 5, if not it will default to 5
        voltage = voltage % 16          # Voltage is represented by a 4-bit number, so max is 21
        self._voltage = int(voltage)
        self._update_VRDV_register()

    def set_current(self, current_type):
        self._hi_current_mode = CURRENT[current_type]
        self._update_VRDV_register()

    def set_io_mode(self, io_mode):
        self._io_mode = IO[io_mode]
        self._update_DC_Register()

    def set_pulse_count(self, pulse_count):
        pulse_count = pulse_count % 64  # Pulse count is a 6-bit number, so max is 63
        self._pulseCount = int(pulse_count)
        self._update_BP_register()

# ----------------------------- Private functions ----------------------------- #

    def _update_VRDV_register(self):
        data = (self._hi_z_mode << 5) + (self._hi_current_mode << 4) + self._voltage
        self._control_register(RW['WRITE'], 0x16, data)

    def _update_BP_register(self):
        data = self._pulseCount
        if self._pre_driver_mode:
            data += 0x40
        self._control_register(RW['WRITE'], 0x1A, data)

    def _update_DC_Register(self):
        data = self._io_mode
        self._control_register(RW['WRITE'], 0x14, data)

    def _control_register(self, rw, address, data):
        address = address & 0x3F;
        comm = rw + (address << 9) + data
        if not self._find_parity16(comm):
            comm = comm + 0x0100
        self._transfer_command(comm);

    def _find_parity16(self, num):
        num = num ^ (num >> 8)
        num = num ^ (num >> 4)
        num = num ^ (num >> 2)
        num = num ^ (num >> 1)
        return num & 1

    def _transfer_command(self, data16):
        data_bytes = bytearray([data16 >> 8, data16 & 0x00FF])
        rcv_bytes = bytearray(2)
        with self._device as spi:
            spi.write_readinto(data_bytes, rcv_bytes)
        rcv16 = int.from_bytes(rcv_bytes, 'big')
        # print(bin(data16))
        # print(bin(rcv16))
        # print()
